server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -0,0 +1,1192 @@
+      if (basePersistDirLock != null) {
+        basePersistDirLock.release();
+        basePersistDirLockChannel.close();
+        basePersistDirLock = null;
+      }
+    }
+    catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private void initializeExecutors()
+  {
+    final int maxPendingPersists = tuningConfig.getMaxPendingPersists();
+
+    if (persistExecutor == null) {
+      // use a blocking single threaded executor to throttle the firehose when write to disk is slow
+      persistExecutor = MoreExecutors.listeningDecorator(
+          Execs.newBlockingSingleThreaded(
+              "appenderator_persist_%d", maxPendingPersists
+          )
+      );
+    }
+    if (mergeExecutor == null) {
+      // use a blocking single threaded executor to throttle the firehose when write to disk is slow
+      mergeExecutor = MoreExecutors.listeningDecorator(
+          Execs.newBlockingSingleThreaded(
+              "appenderator_merge_%d", 1
+          )
+      );
+    }
+  }
+
+  private void shutdownExecutors()
+  {
+    persistExecutor.shutdownNow();
+    mergeExecutor.shutdownNow();
+  }
+
+  private void resetNextFlush()
+  {
+    nextFlush = new DateTime().plus(tuningConfig.getIntermediatePersistPeriod()).getMillis();
+  }
+
+  /**
+   * Populate "sinks" and "sinkTimeline" with committed segments, and announce them with the segmentAnnouncer.
+   *
+   * @return persisted commit metadata
+   */
+  private Object bootstrapSinksFromDisk()
+  {
+    Preconditions.checkState(sinks.isEmpty(), "Already bootstrapped?!");
+
+    final File baseDir = tuningConfig.getBasePersistDirectory();
+    if (!baseDir.exists()) {
+      return null;
+    }
+
+    final File[] files = baseDir.listFiles();
+    if (files == null) {
+      return null;
+    }
+
+    final File commitFile = computeCommitFile();
+    final Committed committed;
+    try {
+      if (commitFile.exists()) {
+        committed = objectMapper.readValue(commitFile, Committed.class);
+      } else {
+        committed = Committed.nil();
+      }
+    }
+    catch (Exception e) {
+      throw new ISE(e, "Failed to read commitFile: %s", commitFile);
+    }
+
+    log.info("Loading sinks from[%s]: %s", baseDir, committed.getHydrants().keySet());
+
+    for (File sinkDir : files) {
+      final File identifierFile = new File(sinkDir, IDENTIFIER_FILE_NAME);
+      if (!identifierFile.isFile()) {
+        // No identifier in this sinkDir; it must not actually be a sink directory. Skip it.
+        continue;
+      }
+
+      try {
+        final SegmentIdentifier identifier = objectMapper.readValue(
+            new File(sinkDir, "identifier.json"),
+            SegmentIdentifier.class
+        );
+
+        final int committedHydrants = committed.getCommittedHydrants(identifier.getIdentifierAsString());
+
+        if (committedHydrants <= 0) {
+          log.info("Removing uncommitted sink at [%s]", sinkDir);
+          FileUtils.deleteDirectory(sinkDir);
```

Comment:
I think we can't, because we might need to re-use this directory later, and it'd be bad if it already has files in it.

I guess we could skip delete failures here, and then try to delete again later on if we are asked to re-use the directory, and then if *that* fails we can defer failure to that point. but it seems better to fail early, in the unlikely event that we cannot write to our own persist dir.